### PR TITLE
Improve the kubelet check error reporting in the output of `agent status`

### PR DESCRIPTION
**Pull request by L3n41c**
### What does this PR do?

In combination with DataDog/integrations-core#7495 , this change improves the `kubelet` check error reported in the output of `agent status` when the agent cannot properly connect to the kubelet.

### Motivation

In case the agent cannot properly connect to the kubelet, the useful details were in the logs but the output of the `agent status` command gave no clue about the reasons.

Here is an example of the `agent status` output in such a case:
```
$ agent status
[…]

=========
Collector
=========

  Running Checks
  ==============

    kubelet (4.1.1)
    ---------------
      Instance ID: kubelet:d884b5186b651429 [ERROR]
      Configuration Source: file:/etc/datadog-agent/conf.d/kubelet.d/conf.yaml.default
      Total Runs: 1
      Metric Samples: Last Run: 0, Total: 0
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 2ms
      Last Execution Date : 2020-09-03 11:40:31.000000 UTC
      Last Successful Execution Date : Never
      Error: Unable to detect the kubelet URL automatically.
      Traceback (most recent call last):
        File "/opt/datadog-agent/embedded/lib/python3.8/site-packages/datadog_checks/base/checks/base.py", line 827, in run
          self.check(instance)
        File "/opt/datadog-agent/embedded/lib/python3.8/site-packages/datadog_checks/kubelet/kubelet.py", line 297, in check
          raise CheckException("Unable to detect the kubelet URL automatically.")
      datadog_checks.base.errors.CheckException: Unable to detect the kubelet URL automatically.
```

Here is what the output becomes with this PR:

```
$ agent status
[…]

=========
Collector
=========

  Running Checks
  ==============

    kubelet (4.1.1)
    ---------------
      Instance ID: kubelet:d884b5186b651429 [ERROR]
      Configuration Source: file:/etc/datadog-agent/conf.d/kubelet.d/conf.yaml.default
      Total Runs: 1
      Metric Samples: Last Run: 0, Total: 0
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 3ms
      Last Execution Date : 2020-09-03 11:14:20.000000 UTC
      Last Successful Execution Date : Never
      Error: Unable to detect the kubelet URL automatically: cannot set a valid kubelet host: cannot connect to kubelet using any of the given hosts: [1.2.3.4] [], Errors: [Get https://1.2.3.4:10250/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers) cannot connect: http: "Get http://1.2.3.4:10255/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)"]
      Traceback (most recent call last):
        File "/opt/datadog-agent/embedded/lib/python3.8/site-packages/datadog_checks/base/checks/base.py", line 827, in run
          self.check(instance)
        File "/opt/datadog-agent/embedded/lib/python3.8/site-packages/datadog_checks/kubelet/kubelet.py", line 297, in check
          raise CheckException("Unable to detect the kubelet URL automatically: " + kubelet_conn_info.get('err'))
      datadog_checks.base.errors.CheckException: Unable to detect the kubelet URL automatically: cannot set a valid kubelet host: cannot connect to kubelet using any of the given hosts: [1.2.3.4] [], Errors: [Get https://1.2.3.4:10250/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers) cannot connect: http: "Get http://1.2.3.4:10255/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)"]
```

### Additional Notes

This would help the investigation of issues like DataDog/integrations-core#2582.

### Describe your test plan

Start the agent in a context where it schedules the `kubelet` check, but it cannot connect to it:
```shell
docker run --rm --name datadog-agent -e DD_API_KEY=$DD_API_KEY -e KUBERNETES=yes -e DD_KUBERNETES_KUBELET_HOST=1.2.3.4 datadog/agent:7.23.0
```

### CODE REVIEW

**Review by ahmed-mez — APPROVED** [review]

File: pkg/collector/python/kubeutil.go [hunk 1]
```diff
@@ -34,9 +35,12 @@ var (
 func getConnections() map[string]string {
 	kubeutil, err := kubelet.GetKubeUtil()
 	if err != nil {
-		// Connection to the kubelet fail, return empty dict
+		// Connection to the kubelet fail, return the error
 		log.Errorf("connection to kubelet failed: %v", err)
-		return nil
+		if e, ok := err.(*retry.Error); ok {
```
Comment (xornivore) [inline pkg/collector/python/kubeutil.go:40]: We could probably use `errors.As` here? https://golang.org/pkg/errors/#example_As

**Review by xornivore — APPROVED** [review]
LGTM just a small comment on using typecast.

File: pkg/collector/python/kubeutil.go (hunk 1 above)
Comment (L3n41c) [reply pkg/collector/python/kubeutil.go]: Good point !
https://github.com/DataDog/datadog-agent/pull/6315/commits/0a3af81224cb89e13882c4e490a2f12e628d397c

**Review by clamoriniere — APPROVED** [review]
I like the usage of `errors.As()` !!! 💯

File: pkg/util/retry/retrier.go [hunk 2]
```diff
@@ -104,10 +105,10 @@ func (r *Retrier) doTry() *Error {
 	}
 	method := r.cfg.AttemptMethod
 	r.RUnlock()
-	err := method()
+	r.lastTryError = method()
```
Comment (olivielpeau) [inline pkg/util/retry/retrier.go:108]: is it safe to write to `r.lastTryError` while the mutex is unlocked?

**Review by olivielpeau — APPROVED** [review]
Small question, other than that LGTM 👍

**Pull request merged**